book-content/chapters/12-the-weird-parts.md

```diff
@@ -0,0 +1,1653 @@
+```
+
+This is like an advanced form of narrowing, where the type of the variable is narrowed based on the value assigned to it.
+
+### Evolving `any` Arrays
+
+This technique of using the evolving `any` also works with arrays. When you declare an array without a specific type, you can push various types of elements to it:
+
+```ts twoslash
+const evolvingArray = [];
+
+evolvingArray.push("abc");
+
+const elem = evolvingArray[0];
+//    ^?
+
+evolvingArray.push(123);
+
+const elem2 = evolvingArray[1];
+//    ^?
+```
+
+Even without specifying types, TypeScript is incredibly smart about picking up on your actions and the behavior you're pushing to evolving `any` types.
+
+## Excess Property Warnings
+
+A deeply confusing part of TypeScript is how it handles excess properties in objects. In many situations, TypeScript won't show errors you might expect when working with objects.
+
+Let's create an `Album` interface that includes `title` and `releaseYear` properties:
+
+```tsx
+interface Album {
+  title: string;
+  releaseYear: number;
+}
+```
+
+Here we create an untyped `rubberSoul` object that includes an excess `label` property:
+
+```tsx
+const rubberSoul = {
+  title: "Rubber Soul",
+  releaseYear: 1965,
+  label: "Parlophone",
+};
+```
+
+Now if we create a `processAlbum` function that accepts an `Album` and logs it, we can pass in the `rubberSoul` object without any issues:
+
+```tsx
+const processAlbum = (album: Album) => console.log(album);
+
+processAlbum(rubberSoul); // No error!
+```
+
+This seems strange! We would expect TypeScript to show an error for the excess `label` property, but it doesn't.
+
+Even more strangely, when we pass the object _inline_, we do get an error:
+
+```ts twoslash
+// @errors: 2353
+type Album = {
+  title: string;
+  releaseYear: number;
+};
+const processAlbum = (album: Album) => console.log(album);
+
+// ---cut---
+processAlbum({
+  title: "Rubber Soul",
+  releaseYear: 1965,
+  label: "Parlophone",
+});
+```
+
+Why the different behavior?
+
+### No Excess Property Checks On Variables
+
+In the first example, we assigned the album to a variable, then passed the variable into our function. In this situation, TypeScript won't check for excess properties.
+
+The reason is that we might be using that variable in other places where the excess property is needed. TypeScript doesn't want to get in the way of that.
+
+But when we inline the object, TypeScript knows that we're not going to use it elsewhere, so it checks for excess properties.
+
+This can make you _think_ that TypeScript cares about excess properties - but it doesn't. It only checks for them in certain situations.
+
+This behavior can be frustrating when you misspell the names of an optional parameter. Imagine you misspell `timeout` as `timeOut`:
+
+```typescript
+const fetch = (options: { timeout?: number }) => {
+  // Implementation
+};
+
+const options = {
+  timeOut: 1000,
+};
+
+fetch(options); // No error!
```

Comment:
this is not a good example because this is actually an error! ([TS playground](https://www.typescriptlang.org/play/?ts=5.6.0-dev.20240721#code/MYewdgzgLgBAZgUysAFjAvDAFCADlAS3AgC4YBvGQgWwRAFcoB+Mse6gIwQCcYBfAJQYAfBQBQMGAHopMAJLVcAGwS0wUAIaFwYvgG4xY0JFh5tkDOMk0EAeUZkAjAAZXAGl0GxiZChz4iSAE9aVkAORAYHm4QbgBCQwQAD1xY2HI+IA))

> Type '{ timeOut: number; }' has no properties in common with type '{ timeout?: number | undefined; }'.(2559)

The parameter type here is a weak type - so whatever we try to pass in there has to have *some* overlap with it.